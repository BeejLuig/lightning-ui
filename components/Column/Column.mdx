import { Preview, Story } from '@storybook/addon-docs/blocks';

# Column

Stuff stuff stuff

<Preview>
  <Story id="column--basic" />
</Preview>