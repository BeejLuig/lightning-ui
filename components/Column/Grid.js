import lng from 'wpe-lightning';
import FocusManager from '../FocusManager/FocusManager';
import { GRID } from '../Styles/Styles';

const BOUNDS = 1000;
export default class Grid extends lng.Component {
  static _template() {
    return {
      boundsMargin: [BOUNDS, BOUNDS, 0, 0],
      Rows: {
        type: FocusManager,
        direction: 'column',
        signals: {
          selectedChange: '_scroll'
        },
        x: GRID.margin.x,
        w: w => w - 2 * GRID.margin.x,
        y: GRID.gutters.vertical
      },
      loading: true
    };
  }

  constructor(...args) {
    super(...args);
    // Ensure setters are called after all properties are set like width
    this._whenEnabled = new Promise(
      resolve =>
        (this._firstEnable = () => {
          this._gridEnabled = true;
          resolve();
        })
    );
  }

  _init() {
    super._init();
    this._originalH = this.h;
  }

  _scroll(selectedRow, prevRow, direction) {
    if (!this.static) {
      if (prevRow && prevRow.currentItem) {
        let index = this._getIndexOfItemNear(selectedRow, prevRow);
        selectedRow._selectedIndex = index;
      }

      if (direction === 'next') {
        this._scrollDown(prevRow);
      } else {
        this._scrollUp();
      }
    }
  }

  _getIndexOfItemNear(selectedRow, prevRow) {
    let prevItem = prevRow.currentItem;
    let [itemX] = prevItem.core.getAbsoluteCoords(-prevRow.offset, 0);
    let index = selectedRow.items.findIndex(item => {
      let [x] = item.core.getAbsoluteCoords(0, 0);
      return x >= itemX || itemX <= x + item.w;
    });

    if (index === -1) {
      return selectedRow.items.length - 1;
    }

    return index;
  }

  scrollTo(index, duration = this.rowTransition.duration * 100) {
    for (let i = 0; i !== Math.abs(this.selectedIndex - index); i++) {
      setTimeout(() => {
        this.selectedIndex > index
          ? this.selectedIndex--
          : this.selectedIndex++;
      }, duration * i);
    }
  }

  _scrollUp() {
    this.render();
  }

  _scrollDown(rowAbove) {
    this.render(rowAbove);

    if (this._nearEnd() && this._getMoreRows) {
      this.provider = this._getMoreRows();
      this._getMoreRows = false;
    }
  }

  _nearEnd(BUFFER = 6) {
    return this.rows.length && this.selectedIndex > this.rows.length - BUFFER;
  }

  _getFocused() {
    return this._Rows;
  }

  resetIndex() {
    this.selectedIndex = 0;
    this.rows && this.rows.forEach(row => (row.selectedIndex = 0));
  }

  set rowsPos(pos) {
    this._Rows.x = pos.x !== undefined ? pos.x : this._Rows.x;
    this._Rows.y = pos.y !== undefined ? pos.y : this._Rows.y;
  }

  get _Rows() {
    return this.tag('Rows');
  }

  get selectedIndex() {
    return this._Rows.selectedIndex;
  }

  set selectedIndex(index) {
    this._Rows.selectedIndex = index;
  }

  get rows() {
    return this._Rows.children;
  }

  set rows(rows) {
    if (this._gridEnabled) {
      this._Rows.childList.clear();
      if (this.fallbacks)
        Object.keys(this.fallbacks).forEach(
          key => (this[`is${[key]}`] = false)
        );
      this.appendRows(rows);
    } else {
      this._whenEnabled.then(() => (this.rows = rows));
    }
  }

  get currentRow() {
    return this.rows[this.selectedIndex];
  }

  set provider(provider) {
    if (provider) {
      provider.then(data => {
        if (Array.isArray(data)) {
          this.rows = data;
        } else {
          data.appendRows
            ? this.appendRows(data.rows)
            : (this.rows = data.rows);
          this._getMoreRows = data.getMoreRows;
        }
      });
    }
  }

  appendRows(rows = []) {
    let rowWidth = this._Rows.renderWidth;
    // Add rows past the bounds margin so they don't load
    let bottomOfScreen = this.y + this.h + 2 * BOUNDS;
    rows.forEach(row => {
      row.w = row.w || rowWidth;
      row.y = bottomOfScreen;
      row.parentGridFocus = this.hasFocus();
      this._Rows.childList.add(this.application.stage.c(row));
    });

    // Ensure rows are drawn so they have height
    this.stage.update();
    this.loading = false;
    this._refocus();
    this.render();
  }

  _computeLastIndex() {
    let totalRows = this.rows.length;
    let MAX_HEIGHT =
      (this.parentGridH ? this.parentGridH : this._originalH) - this.rowSpacing;

    for (let i = totalRows - 1; i >= 0; i--) {
      MAX_HEIGHT -= this.rows[i].h + this.rowSpacing;
      if (MAX_HEIGHT <= 0) {
        return i + 1;
      }
    }

    return 0;
  }

  _patchFallback(state) {
    if (this.fallbacks && this.fallbacks[state]) {
      this.patch({ Rows: { [state]: this.fallbacks[state] } });
      this[`is${[state]}`] = true;
    }
  }

  render(rowAbove) {
    let rowY = 0;
    let index = this.static ? 0 : this.selectedIndex;
    let totalRows = this.rows.length;
    let GRID_HEIGHT = this._originalH + 1.5 * BOUNDS;
    let lastIndex = this._computeLastIndex();

    if (index > lastIndex && this._getMoreRows === undefined) {
      index = lastIndex;
    } else if (rowAbove) {
      rowAbove.smooth = {
        y: [-(rowAbove.h + this.rowSpacing), this.rowTransition],
        alpha: 0
      };
    }

    while (rowY < GRID_HEIGHT && index < totalRows && totalRows) {
      let row = this.rows[index];
      row.smooth = { y: [rowY, this.rowTransition], alpha: 1 };
      rowY += row.h + this.rowSpacing;
      index++;
    }

    if (!this._originalH && rowY) {
      this.h = this._totalH;
      this.fireAncestors('$gridChanged');
    }
  }

  get _totalH() {
    return this.rows.reduce(
      (acc, cur, index) =>
        (acc += cur.h + (index < this.rows.length - 1 ? this.rowSpacing : 0)),
      this._Rows && this._Rows.y ? this._Rows.y : 0
    );
  }

  set rowSpacing(val) {
    this._rowSpacing = val;
    this.render();
  }

  get rowSpacing() {
    return this._rowSpacing || GRID.gutters.horizontal;
  }

  set updateRows(callback) {
    if (this._gridEnabled) {
      this.rows.forEach(callback);
      this.render();
    } else {
      this._whenEnabled.then(() => (this.updateRows = callback));
    }
  }

  get rowTransition() {
    return (
      this._rowTransition || {
        duration: 0.4,
        timingFunction: 'cubic-bezier(0.20, 1.00, 0.30, 1.00)'
      }
    );
  }

  set rowTransition(val) {
    this._rowTransition = val;
  }

  _focus() {
    // Wait till Rows are focused before rendering
    setTimeout(() => this.render(), 0);
    this.rows.forEach(row => (row.parentGridFocus = true));
  }

  _unfocus() {
    this.rows.forEach(row => (row.parentGridFocus = false));
  }

  $removeRow(row) {
    if (row) {
      let wasSelected = row === this._Rows.selected;
      this._Rows.childList.remove(row);
      if (wasSelected || this.selectedIndex >= this.rows.length) {
        this.selectedIndex = this.selectedIndex;
      }

      let rowBefore;
      if (this.selectedIndex > 0) {
        rowBefore = this.rows[this.selectedIndex - 1];
      }
      this.render(rowBefore);
      if (!this.rows.length) {
        this._patchFallback('EmptyState');
      }
    }
  }

  $gridChanged() {
    this.render();
  }
}
