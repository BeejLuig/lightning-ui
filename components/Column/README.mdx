# FocusManager

In Lightning, focus must be explicitly set on an element and managed manually. `FocusManager` is a utility component that simplifies focus for elements in a list.

## Usage

Within a parent component `_template()` function, define a child of type `FocusManager` and in `_getFocused()` return that child component 

```js
import FocusManager from './FocusManager';

class RowExample extends lng.Component {
  static _template() {
    return {
      Row: {
        type: FocusManager,
        children: [
          // components to be managed
        ]
      }
    };
  }
  
  _getFocused() {
      return this.tag('Row');
  }
};
```

By default, `FocusManager` handles left and right key events. You can also explicitly set the `direction` property to `row` for left/right or `column` for up/down.

```js
{
  type: FocusManager,
  direction: 'column',
  children: [/* ... */]
}
```