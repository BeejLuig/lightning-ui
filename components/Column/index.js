import lng from 'wpe-lightning';
import FocusManager from '../FocusManager';
import { COLUMN } from '../Styles/Styles';

const BOUNDS = 1000;
export default class Column extends lng.Component {
  static _template() {
    return {
      boundsMargin: [BOUNDS, BOUNDS, 0, 0],
      Items: {
        type: FocusManager,
        direction: 'column',
        signals: {
          selectedChange: '_scroll'
        },
        x: COLUMN.margin.x,
        w: w => w - 2 * COLUMN.margin.x,
        y: COLUMN.gutters.vertical
      },
      loading: true
    };
  }

  constructor(...args) {
    super(...args);
    // Ensure setters are called after all properties are set like width
    this._whenEnabled = new Promise(
      resolve =>
        (this._firstEnable = () => {
          this._enabled = true;
          resolve();
        })
    );
  }

  _init() {
    super._init();
    this._originalH = this.h;
  }

  _scroll(selected, prev, direction) {
    if (!this.static) {
      if (prev && prev.currentItem) {
        let index = this._getIndexOfItemNear(selected, prev);
        selected._selectedIndex = index;
      }

      if (direction === 'next') {
        this._scrollDown(prev);
      } else {
        this._scrollUp();
      }
    }
  }

  _getIndexOfItemNear(selected, prev) {
    let prevItem = prev.currentItem;
    let [itemX] = prevItem.core.getAbsoluteCoords(-prev.offset, 0);
    let index = selected.items.findIndex(item => {
      let [x] = item.core.getAbsoluteCoords(0, 0);
      return x >= itemX || itemX <= x + item.w;
    });

    if (index === -1) {
      return selected.items.length - 1;
    }

    return index;
  }

  scrollTo(index, duration = this.itemTransition.duration * 100) {
    for (let i = 0; i !== Math.abs(this.selectedIndex - index); i++) {
      setTimeout(() => {
        this.selectedIndex > index
          ? this.selectedIndex--
          : this.selectedIndex++;
      }, duration * i);
    }
  }

  _scrollUp() {
    this.render();
  }

  _scrollDown(itemAbove) {
    this.render(itemAbove);

    if (this._nearEnd() && this._getMoreItems) {
      this.provider = this._getMoreItems();
      this._getMoreItems = false;
    }
  }

  _nearEnd(BUFFER = 6) {
    return this.items.length && this.selectedIndex > this.items.length - BUFFER;
  }

  _getFocused() {
    return this._Items;
  }

  resetIndex() {
    this.selectedIndex = 0;
    this.items && this.items.forEach(item => (item.selectedIndex = 0));
  }

  set itemsPos(pos) {
    this._Items.x = pos.x !== undefined ? pos.x : this._Items.x;
    this._Items.y = pos.y !== undefined ? pos.y : this._Items.y;
  }

  get _Items() {
    return this.tag('Items');
  }

  get selectedIndex() {
    return this._Items.selectedIndex;
  }

  set selectedIndex(index) {
    this._Items.selectedIndex = index;
  }

  get items() {
    return this._Items.children;
  }

  set items(items) {
    if (this._columnEnabled) {
      this._Items.childList.clear();
      if (this.fallbacks)
        Object.keys(this.fallbacks).forEach(
          key => (this[`is${[key]}`] = false)
        );
      this.appendItems(items);
    } else {
      this._whenEnabled.then(() => (this.items = items));
    }
  }

  get currentItem() {
    return this.items[this.selectedIndex];
  }

  set provider(provider) {
    if (provider) {
      provider.then(data => {
        if (Array.isArray(data)) {
          this.items = data;
        } else {
          data.appendItems
            ? this.appendItems(data.items)
            : (this.items = data.items);
          this._getMoreItems = data.getMoreItems;
        }
      });
    }
  }

  appendItems(items = []) {
    let itemWidth = this._Items.renderWidth;
    // Add items past the bounds margin so they don't load
    let bottomOfScreen = this.y + this.h + 2 * BOUNDS;
    items.forEach(item => {
      item.w = item.w || itemWidth;
      item.y = bottomOfScreen;
      item.parentFocus = this.hasFocus();
      this._Items.childList.add(this.application.stage.c(item));
    });

    // Ensure items are drawn so they have height
    this.stage.update();
    this.loading = false;
    this._refocus();
    this.render();
  }

  _computeLastIndex() {
    let totalItems = this.items.length;
    let MAX_HEIGHT =
      (this.parentH ? this.parentH : this._originalH) - this.itemSpacing;

    for (let i = totalItems - 1; i >= 0; i--) {
      MAX_HEIGHT -= this.items[i].h + this.itemSpacing;
      if (MAX_HEIGHT <= 0) {
        return i + 1;
      }
    }

    return 0;
  }

  _patchFallback(state) {
    if (this.fallbacks && this.fallbacks[state]) {
      this.patch({ Items: { [state]: this.fallbacks[state] } });
      this[`is${[state]}`] = true;
    }
  }

  render(itemAbove) {
    let itemY = 0;
    let index = this.static ? 0 : this.selectedIndex;
    let totalItems = this.items.length;
    let COLUMN_HEIGHT = this._originalH + 1.5 * BOUNDS;
    let lastIndex = this._computeLastIndex();

    if (index > lastIndex && this._getMoreItems === undefined) {
      index = lastIndex;
    } else if (itemAbove) {
      itemAbove.smooth = {
        y: [-(itemAbove.h + this.itemSpacing), this.itemTransition],
        alpha: 0
      };
    }

    while (itemY < COLUMN_HEIGHT && index < totalItems && totalItems) {
      let item = this.items[index];
      item.smooth = { y: [itemY, this.itemTransition], alpha: 1 };
      itemY += item.h + this.itemSpacing;
      index++;
    }

    if (!this._originalH && itemY) {
      this.h = this._totalH;
      this.fireAncestors('$columnChanged');
    }
  }

  get _totalH() {
    return this.items.reduce(
      (acc, cur, index) =>
        (acc += cur.h + (index < this.items.length - 1 ? this.itemSpacing : 0)),
      this._Items && this._Items.y ? this._Items.y : 0
    );
  }

  set itemSpacing(val) {
    this._itemSpacing = val;
    this.render();
  }

  get itemSpacing() {
    return this._itemSpacing || COLUMN.gutters.horizontal;
  }

  set updateItems(callback) {
    if (this._columnEnabled) {
      this.items.forEach(callback);
      this.render();
    } else {
      this._whenEnabled.then(() => (this.updateItems = callback));
    }
  }

  get itemTransition() {
    return (
      this._itemTransition || {
        duration: 0.4,
        timingFunction: 'cubic-bezier(0.20, 1.00, 0.30, 1.00)'
      }
    );
  }

  set itemTransition(val) {
    this._itemTransition = val;
  }

  _focus() {
    // Wait till Items are focused before rendering
    setTimeout(() => this.render(), 0);
    this.items.forEach(item => (item.parentFocus = true));
  }

  _unfocus() {
    this.items.forEach(item => (item.parentFocus = false));
  }

  $removeItem(item) {
    if (item) {
      let wasSelected = item === this._Items.selected;
      this._Items.childList.remove(item);
      if (wasSelected || this.selectedIndex >= this.items.length) {
        this.selectedIndex = this.selectedIndex;
      }

      let itemBefore;
      if (this.selectedIndex > 0) {
        itemBefore = this.items[this.selectedIndex - 1];
      }
      this.render(itemBefore);
      if (!this.items.length) {
        this._patchFallback('EmptyState');
      }
    }
  }

  $columnChanged() {
    this.render();
  }
}
