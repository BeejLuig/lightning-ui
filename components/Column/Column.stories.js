import lng from 'wpe-lightning';
import Column from '.';
import FocusManager from '../FocusManager';

export default {
  title: 'Column',
  component: Column
};

export const Basic = () =>
  class BasicExample extends lng.Component {
    static _template() {
      return {
        Column: {
          type: Column,
          children: Array.apply(null, { length: 5 }).map((_, i) => ({
            type: Button,
            buttonText: 'Button',
            y: i * 50
          }))
        }
      };
    }

    _getFocused() {
      return this.tag('Column');
    }
  };

export const Scrolling = () =>
  class ScrollingExample extends lng.Component {
    static _template() {
      return {
        Column: {
          type: Column,
          children: Array.apply(null, { length: 20 }).map((_, i) => ({
            type: Button,
            buttonText: `Button ${i + 1}`,
            y: i * 50
          }))
        }
      };
    }

    _getFocused() {
      return this.tag('Column');
    }
  };

export const MultiColumn = () =>
  class MultiColumnExample extends lng.Component {
    static _template() {
      return {
        FocusManager: {
          type: FocusManager,
          direction: 'row',
          children: [{ type: Basic() }, { type: Basic(), x: 160 }]
        }
      };
    }

    _getFocused() {
      return this.tag('FocusManager');
    }
  };

export const SkipFocus = () =>
  class ColumnExample extends lng.Component {
    static _template() {
      return {
        x: 20,
        y: 20,
        Column: {
          type: FocusManager,
          direction: 'column',
          children: Array.apply(null, { length: 50 }).map((_, i) => {
            if (i % 4 === 0)
              return {
                type: Title,
                titleText: 'Title',
                y: i * 50,
                skipFocus: true
              };
            return { type: Button, buttonText: 'Button', y: i * 50 };
          })
        }
      };
    }

    _getFocused() {
      return this.tag('Column');
    }
  };

class Title extends lng.Component {
  static _template() {
    return {
      Label: {
        x: 75,
        y: 22,
        mount: 0.5,
        color: 0xffffffff,
        text: { fontSize: 20 }
      }
    };
  }

  _init() {
    this.tag('Label').text = this.titleText;
  }
}

class Button extends lng.Component {
  static _template() {
    return {
      color: 0xff1f1f1f,
      texture: lng.Tools.getRoundRect(150, 40, 4),
      Label: {
        x: 75,
        y: 22,
        mount: 0.5,
        color: 0xffffffff,
        text: { fontSize: 20 }
      }
    };
  }
  _init() {
    this.tag('Label').text = this.buttonText;
  }
  _focus() {
    this.color = 0xffffffff;
    this.tag('Label').color = 0xff1f1f1f;
  }
  _unfocus() {
    this.color = 0xff1f1f1f;
    this.tag('Label').color = 0xffffffff;
  }
}
