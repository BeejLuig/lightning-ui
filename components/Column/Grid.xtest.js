import Grid from 'src/components/Grid';
import Row from 'src/components/Row';
import TestRenderer from 'test/lightning-test-renderer';
import lng from 'wpe-lightning';

const baseItem = {
  type: lng.Component,
  w: 80,
  h: 80
};
const items = [
  { ...baseItem },
  { ...baseItem },
  { ...baseItem },
  { ...baseItem },
  { ...baseItem }
];

const baseRow = {
  type: Row,
  h: 80,
  items
};

const rows = [
  { ...baseRow },
  { ...baseRow },
  { ...baseRow },
  { ...baseRow },
  { ...baseRow }
];

const Component = {
  Component: {
    type: Grid,
    title: 'My Grid',
    h: 800,
    rowTransition: { duration: 0 },
    rowSpacing: 80,
    rows
  }
};

describe('Grid', () => {
  let testRenderer, grid;

  beforeEach(() => {
    testRenderer = TestRenderer.create(Component);
    grid = testRenderer.getInstance();
    return grid._whenEnabled;
  });

  it('should render', () => {
    let tree = testRenderer.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should set focus on first row', () => {
    expect(grid.rows[0].hasFocus()).toBe(true);
  });

  it('should set focus on next row on keyDown', () => {
    testRenderer.keyPress('Down');
    expect(grid.rows[1].hasFocus()).toBe(true);
  });

  describe('rowSpacing', () => {
    it('should set spacing', () => {
      let spacing = 100;
      let row = grid.rows[1];
      grid.rowSpacing = spacing;
      testRenderer.update();
      expect(row.y).toBe(spacing + row.h);
    });
  });

  describe('provider', () => {
    it('should take a promise to append items', done => {
      grid.provider = Promise.resolve({
        appendRows: true,
        rows: [{ ...baseRow }, { ...baseRow }]
      });

      setTimeout(() => {
        expect(grid.rows.length).toBe(7);
        done();
      }, 1);
    });

    it('should take a promise to replace items', done => {
      grid.provider = Promise.resolve({
        rows: [{ ...baseRow }, { ...baseRow }]
      });

      setTimeout(() => {
        expect(grid.rows.length).toBe(2);
        done();
      }, 1);
    });
  });

  describe('listeners', () => {
    it('should listen for $removeRow', () => {
      let row = grid.rows[1];
      grid.$removeRow(row);
      testRenderer.update();
      expect(grid.rows.length).toBe(4);
    });
  });

  describe('scrolling', () => {
    it('should scroll down', () => {
      let row = grid.rows[0];
      testRenderer.keyPress('Down');
      testRenderer.keyPress('Down');
      testRenderer.keyPress('Down');
      testRenderer.keyPress('Down');
      testRenderer.update();
      expect(row.y).toBe(-160);
    });

    it('should scroll up', () => {
      let row = grid.rows[0];
      testRenderer.keyPress('Down');
      testRenderer.keyPress('Up');
      testRenderer.update();
      expect(row.y).toBe(0);
    });

    it('should scroll to index before', () => {
      jest.useFakeTimers();
      grid.selectedIndex = 3;
      grid.scrollTo(0);
      jest.runAllTimers();
      expect(grid.selectedIndex).toBe(0);
    });

    it('should scroll to index after', () => {
      jest.useFakeTimers();
      grid.selectedIndex = 0;
      grid.scrollTo(3);
      jest.runAllTimers();
      expect(grid.selectedIndex).toBe(3);
    });

    it('should set selected item for row based on previous row', () => {
      let row = grid.rows[0];
      row._selectedIndex = 3;
      testRenderer.update();
      testRenderer.keyPress('Down');
      testRenderer.update();
      expect(grid.rows[1]._selectedIndex).toBe(3);
    });

    it('should keep a full screen of rows', () => {
      let row = grid.rows[1];
      testRenderer.keyPress('Down');
      testRenderer.keyPress('Down');
      testRenderer.keyPress('Down');
      testRenderer.keyPress('Down');
      testRenderer.update();
      expect(row.y).toBe(0);
    });

    it('should load more items near bottom with getMoreRows', () => {
      let mock = jest.fn();
      mock.mockReturnValue(Promise.resolve([]));
      grid._getMoreRows = mock;
      testRenderer.keyPress('Down');
      testRenderer.keyPress('Down');
      expect(mock).toHaveBeenCalled();
    });
  });
});
