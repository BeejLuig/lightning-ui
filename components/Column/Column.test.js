import Column from '.';
import Row from '../Row';
import TestRenderer from '../lightning-test-renderer';
import lng from 'wpe-lightning';

const baseItem = {
  type: lng.Component,
  w: 80,
  h: 80
};

const items = [
  { ...baseItem },
  { ...baseItem },
  { ...baseItem },
  { ...baseItem },
  { ...baseItem }
];

const baseRow = {
  type: Row,
  h: 80,
  items
};

const rows = [
  { ...baseRow },
  { ...baseRow },
  { ...baseRow },
  { ...baseRow },
  { ...baseRow }
];

const Component = {
  Component: {
    type: Column,
    title: 'My Column',
    h: 800,
    itemTransition: { duration: 0 },
    itemSpacing: 80,
    items: rows
  }
};

describe('Column', () => {
  let testRenderer, column;

  beforeEach(() => {
    testRenderer = TestRenderer.create(Component);
    column = testRenderer.getInstance();
    return column._whenEnabled;
  });

  it('should render', () => {
    let tree = testRenderer.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should set focus on first row', () => {
    expect(column.rows[0].hasFocus()).toBe(true);
  });

  it('should set focus on next row on keyDown', () => {
    testRenderer.keyPress('Down');
    expect(column.rows[1].hasFocus()).toBe(true);
  });

  describe('rowSpacing', () => {
    it('should set spacing', () => {
      let spacing = 100;
      let row = column.rows[1];
      column.rowSpacing = spacing;
      testRenderer.update();
      expect(row.y).toBe(spacing + row.h);
    });
  });

  describe('provider', () => {
    it('should take a promise to append items', done => {
      column.provider = Promise.resolve({
        appendRows: true,
        rows: [{ ...baseRow }, { ...baseRow }]
      });

      setTimeout(() => {
        expect(column.rows.length).toBe(7);
        done();
      }, 1);
    });

    it('should take a promise to replace items', done => {
      column.provider = Promise.resolve({
        rows: [{ ...baseRow }, { ...baseRow }]
      });

      setTimeout(() => {
        expect(column.rows.length).toBe(2);
        done();
      }, 1);
    });
  });

  describe('listeners', () => {
    it('should listen for $removeRow', () => {
      let row = column.rows[1];
      column.$removeRow(row);
      testRenderer.update();
      expect(column.rows.length).toBe(4);
    });
  });

  describe('scrolling', () => {
    it('should scroll down', () => {
      let row = column.rows[0];
      testRenderer.keyPress('Down');
      testRenderer.keyPress('Down');
      testRenderer.keyPress('Down');
      testRenderer.keyPress('Down');
      testRenderer.update();
      expect(row.y).toBe(-160);
    });

    it('should scroll up', () => {
      let row = column.rows[0];
      testRenderer.keyPress('Down');
      testRenderer.keyPress('Up');
      testRenderer.update();
      expect(row.y).toBe(0);
    });

    it('should scroll to index before', () => {
      jest.useFakeTimers();
      column.selectedIndex = 3;
      column.scrollTo(0);
      jest.runAllTimers();
      expect(column.selectedIndex).toBe(0);
    });

    it('should scroll to index after', () => {
      jest.useFakeTimers();
      column.selectedIndex = 0;
      column.scrollTo(3);
      jest.runAllTimers();
      expect(column.selectedIndex).toBe(3);
    });

    it('should set selected item for row based on previous row', () => {
      let row = column.rows[0];
      row._selectedIndex = 3;
      testRenderer.update();
      testRenderer.keyPress('Down');
      testRenderer.update();
      expect(column.rows[1]._selectedIndex).toBe(3);
    });

    it('should keep a full screen of rows', () => {
      let row = column.rows[1];
      testRenderer.keyPress('Down');
      testRenderer.keyPress('Down');
      testRenderer.keyPress('Down');
      testRenderer.keyPress('Down');
      testRenderer.update();
      expect(row.y).toBe(0);
    });

    it('should load more items near bottom with getMoreRows', () => {
      let mock = jest.fn();
      mock.mockReturnValue(Promise.resolve([]));
      column._getMoreRows = mock;
      testRenderer.keyPress('Down');
      testRenderer.keyPress('Down');
      expect(mock).toHaveBeenCalled();
    });
  });
});
