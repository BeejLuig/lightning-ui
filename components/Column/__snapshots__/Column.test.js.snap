// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Column should render 1`] = `
Object {
  "Component": Object {
    "Element-3": Object {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "children": Object {
        "Items": Object {
          "active": false,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "children": Object {
            "Element-18": Object {
              "active": false,
              "alpha": 1,
              "attached": true,
              "boundsMargin": null,
              "clipping": false,
              "color": 4294967295,
              "enabled": true,
              "flex": false,
              "flexItem": false,
              "h": 80,
              "hasFinalFocus": true,
              "hasFocus": true,
              "id": 18,
              "isComponent": true,
              "mount": 0,
              "mountX": 0,
              "mountY": 0,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": null,
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": "",
              "tag": [Function],
              "type": "Component",
              "visible": true,
              "w": 80,
              "x": 0,
              "y": 0,
              "zIndex": 0,
            },
            "Element-19": Object {
              "active": false,
              "alpha": 1,
              "attached": true,
              "boundsMargin": null,
              "clipping": false,
              "color": 4294967295,
              "enabled": true,
              "flex": false,
              "flexItem": false,
              "h": 80,
              "hasFinalFocus": false,
              "hasFocus": false,
              "id": 19,
              "isComponent": true,
              "mount": 0,
              "mountX": 0,
              "mountY": 0,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": null,
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": "",
              "tag": [Function],
              "type": "Component",
              "visible": true,
              "w": 80,
              "x": 120,
              "y": 0,
              "zIndex": 0,
            },
            "Element-20": Object {
              "active": false,
              "alpha": 1,
              "attached": true,
              "boundsMargin": null,
              "clipping": false,
              "color": 4294967295,
              "enabled": true,
              "flex": false,
              "flexItem": false,
              "h": 80,
              "hasFinalFocus": false,
              "hasFocus": false,
              "id": 20,
              "isComponent": true,
              "mount": 0,
              "mountX": 0,
              "mountY": 0,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": null,
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": "",
              "tag": [Function],
              "type": "Component",
              "visible": true,
              "w": 80,
              "x": 240,
              "y": 0,
              "zIndex": 0,
            },
            "Element-21": Object {
              "active": false,
              "alpha": 1,
              "attached": true,
              "boundsMargin": null,
              "clipping": false,
              "color": 4294967295,
              "enabled": true,
              "flex": false,
              "flexItem": false,
              "h": 80,
              "hasFinalFocus": false,
              "hasFocus": false,
              "id": 21,
              "isComponent": true,
              "mount": 0,
              "mountX": 0,
              "mountY": 0,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": null,
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": "",
              "tag": [Function],
              "type": "Component",
              "visible": true,
              "w": 80,
              "x": 360,
              "y": 0,
              "zIndex": 0,
            },
            "Element-22": Object {
              "active": false,
              "alpha": 1,
              "attached": true,
              "boundsMargin": null,
              "clipping": false,
              "color": 4294967295,
              "enabled": true,
              "flex": false,
              "flexItem": false,
              "h": 80,
              "hasFinalFocus": false,
              "hasFocus": false,
              "id": 22,
              "isComponent": true,
              "mount": 0,
              "mountX": 0,
              "mountY": 0,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": null,
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": "",
              "tag": [Function],
              "type": "Component",
              "visible": true,
              "w": 80,
              "x": 480,
              "y": 0,
              "zIndex": 0,
            },
          },
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 0,
          "hasFinalFocus": false,
          "hasFocus": true,
          "id": 5,
          "isComponent": true,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Items",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "Row",
          "tag": [Function],
          "tags": Array [
            "Items",
          ],
          "type": "Row",
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
        "Title": Object {
          "active": false,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 0,
          "id": 4,
          "isComponent": undefined,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Title",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "tags": Array [
            "Title",
          ],
          "texture": Object {
            "lineHeight": 80,
            "type": "TextTexture",
          },
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
      },
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 80,
      "hasFinalFocus": false,
      "hasFocus": true,
      "id": 3,
      "isComponent": true,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": null,
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": "",
      "tag": [Function],
      "type": "Row",
      "visible": true,
      "w": 560,
      "x": 0,
      "y": 0,
      "zIndex": 0,
    },
  },
}
`;
